py-pip: add 23.1.2

manuelakuhn opened the pull request:
https://github.com/pypa/pip/tree/23.1.2

manuelakuhn committed:
py-pip: add 23.1.2

manuelakuhn committed:
Restrict py-pip version for py-protobuf

manuelakuhn commented:
It seems that `py-protobuf` does not work with the newer `py-pip`, restricting it to `py-pip@:23.0` works again, so I have added this.

I also tested the newer `py-protobuf` version and this works with `py-pip@23.1.2` -> I will open an additional PR for `py-protobuf@4.23.3`

tldahlgren reviewed (commented):
FYI. Confirmed the py-pip version sha256.

manuelakuhn commented:
ok, now we get the same error we got for `py-protobuf` before for `py-xgboost`. I am not sure if restricting the `py-pip` versions is the right way. The error message is not really helpful. Any suggestions

adamjstewart commented:
Looks like `--install-option` was finally removed in 23.1 ([PR](https://github.com/pypa/pip/pull/11858), [Changelog](https://pip.pypa.io/en/stable/news/#v23-1)).

`--install-option` and `--global-option` have been deprecated since `--config-settings` was added to replace them in pip 22.1. However, I still haven't figured out how to use the new flag. I'm not sure if it works with packages that still use legacy setup.py files or whether it only works with proper PEP 621 backends. Maybe @pradyunsg can comment.

Either way, I'm fine with limiting protobuf to older pip when `+cpp`

manuelakuhn committed:
Merge branch 'develop' into py-pip-23.1.2

manuelakuhn committed:
Restrict py-pip version for straightforward packages

manuelakuhn commented:
I checked which packages also use `--install-option` and did not build without a restriction of `py-pip`.

Straightforward ones were:
- py-brian2
- py-cmake
- py-fastrlock
- py-lightgbm+mpi
- py-mpi4py
- py-pyarrow (only builds with +parquet) 
- py-qiskit-aer
- py-scs (variants)
- py-symengine
- py-xgboost

These I have added directly in this PR.

The following I had to modify because they did not build with `--test=root`:
- py-astropy (missing test dependencies; create 4 + update 1 packages)
- py-minkowskiengine (missing openblas dependency, python@3.10 compatibility)
- py-tomopy (straightforward but depends on py-astropy)

Should I add them here as well or open separate PRs? 
For py-abipy I opened a separate PR because I also updated it (#38716).

Packages that don't build for me:
- py-charm4py
- py-networkit
- py-pymol
- py-pyside
- py-pyside2
- py-shiboken
- py-gluoncv (building works but the tests don't)

How should we handle these? Just add the py-pip restriction and ignore that they don't build?

adamjstewart commented:
> Should I add them here as well or open separate PRs?

Separate please

> How should we handle these? Just add the py-pip restriction and ignore that they don't build?

I'm fine with that. If you want to be a hero you can try to get them to build, but that's only if you have the time.

adamjstewart commented on var/spack/repos/builtin/packages/py-brian2/package.py:
```diff
@@ -22,6 +22,9 @@ class PyBrian2(PythonPackage):
     depends_on("python@2.7:", type=("build", "run"))
     depends_on("python@3.6:", type=("build", "run"), when="@2.4:")
     depends_on("python@3.7:", type=("build", "run"), when="@2.5:")
+    # in newer pip versions --install-options does not exist
```
Technically the flag name is `--install-option` singular

manuelakuhn replied:
I called it `--install-options` because its called like that in the [Changelog](https://pip.pypa.io/en/stable/news/#v23-1) and the [Issue](https://github.com/pypa/pip/issues/11358). This is changed now.

adamjstewart commented on var/spack/repos/builtin/packages/py-scs/package.py:
```diff
@@ -31,6 +31,12 @@ class PyScs(PythonPackage, CudaPackage):
+    depends_on("py-pip@:23.0", when="+float32", type="build")
+    depends_on("py-pip@:23.0", when="+extra_verbose", type="build")
+    depends_on("py-pip@:23.0", when="+int32", type="build")
+    depends_on("py-pip@:23.0", when="+blas64", type="build")
```
Honestly I'd be fine with always having this requirement just to make the package simpler.

manuelakuhn replied:
Done.

adamjstewart reviewed (dismissed):
Feel free to merge, the above comments are minor

manuelakuhn committed:
Restrict py-pip version for nrm

manuelakuhn committed:
Fix --install-option name in comments

manuelakuhn committed:
Simplify py-pip restriction for py-scs

manuelakuhn commented:
I also had restrict the `py-pip` version for `nrm` for the pipeline to run through because it does not have a `pyproject.toml` file.
With cases like that, I see quite some packages breaking once this PR is merged.

adamjstewart commented on var/spack/repos/builtin/packages/nrm/package.py:
```diff
@@ -14,6 +14,8 @@ class Nrm(PythonPackage):
     version("0.1.0", sha256="911a848042fa50ed216c818e0667bcd3e4219687eb5a35476b7313abe12106dc")
 
     depends_on("py-setuptools", type=("build"))
+    # pip 23.1 enforces to have a pyproject.toml
```
I find this hard to believe. What was the error message?

manuelakuhn replied:
Actually I just tested it with another package without `pyproject.toml` and noticed that this does not seem to be the problem. I could not make sense out of the error message and when I googled I found somewhere this is due to missing pyproject.toml files. Seems that is wrong. Will try to find out what the actual problem is.

manuelakuhn replied:
Ok, I it seems that there was an error message that I overlooked:
```
    File "$spack/opt/spack/linux-fedora37-x86_64/gcc-12.3.1/python-3.8.17-bl3amrqv7
vomgl6dxgbbpnsllpcg7p6/lib/python3.8/zipfile.py", line 362, in __init__
      raise ValueError('ZIP does not support timestamps before 1980')
  ValueError: ZIP does not support timestamps before 1980
```

manuelakuhn replied:
I have no idea why this happens with py-pip@23.1 but not with py-pip@32.0.
I have exchanged the comment with a description of the error message. Is this enough?

adamjstewart replied:
Yeah, that's good enough for now. Someone who really cares about nrm can fix it.

manuelakuhn committed:
nrm: fix wrong comment
